File: Programmers/Level 3/양과 늑대.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int cnt = 0;

void dfs(int now, int w, int s, vector<int> nextNode, vector<int> info, vector<vector<int>> tree) {
	cnt = max(cnt, s);
	if (w >= s) {
		return;
	}

	for (int i = 0; i < nextNode.size(); i++) {
		vector<int> next = nextNode;
		next.erase(next.begin() + i);
		for (int j = 0; j < tree[nextNode[i]].size(); j++) {
			next.push_back(tree[nextNode[i]][j]);
		}
		if (!info[nextNode[i]]) {
			dfs(nextNode[i], w, s + 1, next, info, tree);
		}
		else {
			dfs(nextNode[i], w + 1, s, next, info, tree);
		}
	}
}

int solution(vector<int> info, vector<vector<int>> edges) {
	vector<vector<int>> tree(info.size());
	for (int i = 0; i < edges.size(); i++) {
		tree[edges[i][0]].push_back(edges[i][1]);
	}

	vector<int> nextNode;
	for (int i = 0; i < tree[0].size(); i++) {
		nextNode.push_back(tree[0][i]);
	}

	dfs(0, 0, 1, nextNode, info, tree);
	return cnt;
}

int main(void) {
	vector<int> info = { 0,0,1,1,1,0,1,0,1,0,1,1 };
	vector<vector<int>> edges = { {0, 1}, {1, 2}, {1, 4}, {0, 8}, {8, 7}, {9, 10}, {9, 11}, {4, 3}, {6, 5}, {4, 6}, {8, 9} };
	int result = solution(info, edges);
	cout << result << "\n";
	return 0;
}

File: Programmers/Level 3/양과늑대.java
import java.util.*;

public class 양과늑대 {
    static int maxCnt = 0;
    static Map<Integer, List<Integer>> tree;
    public static void dfs(int now, int s, int w, int[] info, List<Integer> list){
        if(info[now] == 0) {
            s+=1;
        }
        else {
            w+=1;
        }
        if(w >= s) {
            return;
        }
        maxCnt = Math.max(s, maxCnt);

        List<Integer> next = new ArrayList<>();
        next.addAll(list);
        if(tree.containsKey(now)) {
            next.addAll(tree.get(now));
        }
        next.remove(Integer.valueOf(now));
        for (Integer i : next) {
            dfs(i, s, w, info, next);
        }
    }
    public static int solution(int[] info, int[][] edges) {
        tree = new HashMap<>();
        for (int[] edge : edges) {
            if(!tree.containsKey(edge[0])) {
                tree.put(edge[0], new ArrayList<>());
            }
            tree.get(edge[0]).add(edge[1]);
        }
        List<Integer> list = new ArrayList<>();
        list.add(0);
        dfs(0, 0, 0, info, list);
        return maxCnt;
    }
    public static void main(String[] args) {
        int[] info = {0,0,1,1,1,0,1,0,1,0,1,1};
        int[][] edges = {{0, 1}, {1, 2}, {1, 4}, {0, 8}, {8, 7}, {9, 10}, {9, 11}, {4, 3}, {6, 5}, {4, 6}, {8, 9}};
        int result = solution(info, edges);
        System.out.println("result = " + result);
    }
}
